### DISCUSSION TITLE
Modelable child component not working in loop

### DISCUSSION THREAD

**Discussion by arthurshlain** [livewire/livewire#6955] [7. Bugs (archived)]
### Livewire version

v3.0.5

### Which PHP version are you using?

PHP 8.1

### Repository URL

https://github.com/arthurshlain/livewire-test

### Steps To Reproduce

1. Create form component:

`php artisan make:livewire ItemUpload`

```php
class ItemUpload extends Component
{
    public array $items = [
    [
            'name' => 'Test Item',
            'categories' => []
        ]
    ];

    public array $defaultCategories = [];

    public function render()
    {
        return view('livewire.item-upload');
    }

    public function removeUploadItem($index)
    {
        unset($this->items[$index]);
    }
}
```

```blade
<div>

    <label>Default category</label>
    <livewire:category-multi-select wire:model="defaultCategories"/>
  
    @foreach ($items as $index => $item)
        <div wire:key="{{ $index }}">
            <button aria-label="Remove item" 
                    wire:click.stop="removeUploadItem({{ $index }})">×</button>

            {{-- Item name --}}
            <div>
              <label for="items.{{ $index }}.term_name">Enter item name</label>
              <input placeholder="Enter item name"
                     wire:model="items.{{ $index }}.name" 
                     id="items.{{ $index }}.term_name"/>
            </div>
            
            {{-- Category --}}
            <div>
                <livewire:category-multi-select wire:model="items.{{ $index }}.categories" />
            </div>

        </div>
    @endforeach

</div>
```

2. Create category-multi-select component:

`php artisan make:livewire CategoryMultiSelect`

```php
class CategoryMultiSelect extends Component
{
    #[Modelable]
    public mixed $categories = [];

    public string $label = '';

    public function render():
    {
        return view('livewire.category-multi-select');
    }

    /*public function addCategory($categoryName)
    {
        $category = Category::firstOrCreate(
            ['name' => $categoryName],
            ['slug' => Str::of($categoryName)->slug()]
        );

        return [
            'id'   => $category->id,
            'name' => $category->name
        ];
    }*/

    /*public function selectCategory($category_id)
    {
        $this->categories[] = $category_id;
    }*/
}
```

```blade
<div>
    @if($label)
        <label>{{ $label }}</label>
    @endif
    <select wire:model="categories" multiple>
        {{--@foreach(\App\Models\Category::all() as $category)
            <option value="{{ $category->id }}">{{ $category->name }}</option>
        @endforeach--}}
        <option value="1">Example option 1</option>
        <option value="2">Example option 2</option>
        <option value="3">Example option 3</option>
    </select>
</div>
```

In form above the Default category component is working OK.

But all components in foreach loop not working as expected. I trying a lot of different dot + indexer notations, specify keys etc.

Error in console:

```
livewire.js?id=f41737f6:1227 Alpine Expression Error: Unexpected number

Expression: "$wire.$parent.items.0.categories"
```

Errors dissapear when I remove #[Modelable] attribute from child component, but without modelable the parent form will not be updated.

Replacing this multiselect to blade component like `<x-category-multiselect>` is useless, bacause original idea is to extend this select with livewire method `addCategory` when none of the existing categories fit. When using a blade component, it would be necessary to move the logic for creating a new category into API or something. What is the convenience then?

### Additional information

The native select was used for example. Original question with WireUI select here: https://github.com/livewire/livewire/discussions/6931

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by dmikam** [comment]
Same problem here. Livewire version 3.0.7 - fresh out the oven ;) . But what is worse is that it looks that it changes the syntax to [index]:
```
// Doesn't work for me
wire:model="items.{{ $index }}.categories"
// but this does:
wire:model="items[{{ $index }}].categories"
```

Meanwhile, the inputs and other non #[Modelable] components does not support this "new" syntax. 
You just can't use two different incompatible syntaxes for the same thing.

**Reply by dmikam** [reply]
I'ts even more strange than I thought initially.
Reactivity works in different ways depending on the direction of the reactivity:

1) Passing value and its changes from outside to the component with the #[Modelable] property - works in object.collection.index format and does NOT work in object.collection[index] format
2) Passing the #[Modelable] changes out of the component - works in object.collection[index] format and does NOT work in object.collection.index format.
In other words, both things work in opposite ways.

**Comment by CarlEdmondson** [comment]
I'm also running into this issue.

It appears that the issue can somewhat be fixed by updating the render method in the src/Features/SupportWireModelingNestedComponents/SupportWireModelingNestedComponents class.

If you see below, I've added some code to walk through the variable name and replace any numeric parts of the dot notion with [ ] version. This appears to fix the issue I'm experiencing

    public function render($view, $data)
    {
        return function ($html, $replaceHtml) {
            $bindings = store($this->component)->get('bindings', false);

            if (! $bindings) return;

            // Currently we can only support a single wire:model bound value,
            // so we'll just get the first one. But in the future we will
            // likely want to support named bindings, so we'll keep
            // this value as an array.
            $outer = array_keys($bindings)[0];
            $inner = array_values($bindings)[0];

            // We need to look through the dot notation parts of the outer
            // variable name and check for any numerical values and replace
            // them with [{num}].
            $outer = trim(
                array_reduce(
                    explode(".", $outer),
                    function($carry, $part){ return $carry.(is_numeric($part) ? "[$part]" : ".$part"); },
                    ''
                ),
                '.'
            );

            // Attach the necessary Alpine directives so that the child and
            // parent's JS, ephemeral, values are bound.
            $replaceHtml(Utils::insertAttributesIntoHtmlRoot($html, [
                'x-model' => '$wire.$parent.'.$outer,
                'x-modelable' => '$wire.'.$inner,
            ]));
        };
    }

**Reply by CarlEdmondson** [reply]
I've submitted a pull request #7115 with my changes above, so hopefully this resolves the issue and can be rolled into the next update
